lib/ansible/module_utils/common/text/converters.py

```diff
@@ -25,6 +26,8 @@ def _json_encode_fallback(obj):
         return list(obj)
     elif isinstance(obj, datetime.datetime):
         return obj.isoformat()
+    elif isinstance(obj, Decimal):
+        return float(obj)
```

Comment:
This is incorrect for many usages of Decimal, if the specific module usage does not care about Decimal/float distinctions, it can do the conversion itself before dumping to JSON. Otherwise this introduces a 'trap' for other modules that might try to pass Decimal values and get 'incorrectly' floats instead.

Reply:
@bcoca , you suggest not to handle Decimal at all and leave it as it was, right?

AFAIK, Decimal can be converted to int, float or string by embedded methods.
To return values like '1.0001' as a string is very strange because in playbooks we usually use numeric return values for comparison with other numeric values or arithmetic operations.
I think the most of developers will use float for this.
In another case we must implement our own methods like __convert(self, val) here [#55434](https://github.com/ansible/ansible/pull/55434/files) line 262 and use it many times like in the example.

I suggest to convert Decimal to float by default.

Reply:
@samdoran , what's your opinion?

Reply:
I feel like using `float()` does the right thing. Given this example:

```python
data = {'number': Decimal(1.557)}
json_str_encoded = json.dumps(data, cls=AnsibleJSONEncoder)       # '{"number": 1.557}'
json_str_decoded = json.loads(json_str, cls=AnsibleJSONDecoder)    # {'number': 1.557}
```

Changing a `Decimal` object to a string seems incorrect: `'{"number": "1.557"}`.

I see that we don't actually get a `Decimal` object back when decoding again, so that is lost in translation. I think that is ok since we get a float, not a string, which would be more appropriate when the original data container a numeric value. It _is_ different — `float` != `Decimal` — but for the rare cases where a `Decimal` object is present in the input data, converting to `float` is better than raising a `TypeError` or converting to a string.

Reply:
i disagree, a string will still keep the 'correct' value and can be converted later at user's leisure, while a float will result in information loss.

also note that most variables are stirngified on Ansible side anyways so keeping as a float during jsonisation is not a guarantee when later used in a template that you won't have to still convert from string.

I still insist that this should just be solved in the module itself vs trying to enforce a general solution for all modules.

Reply:
@bcoca , thank you for the review and explanation. Sure, if there are any doubts, this PR should be closed.
It’s difficult to discuss with “general solution”.
Moreover, I agree, it’s easy to implement in modules. So I’ll close it. @sdoran , @bcoca , thank you

Reply:
@bcoca , thank you for the review and explanation. Sure, if there are any doubts, this PR should be closed.
It’s difficult to discuss with “general solution”.
Moreover, I agree, it’s easy to implement in modules. So I’ll close it. @sdoran , @bcoca , thank you